site/documentation/content/concepts/device-identity.md

```diff
@@ -147,3 +147,34 @@ After having verified the client certificate using the trust anchor(s), the prot
 the client certificate's *subject DN* and invokes the Credentials API's
 [*get Credentials*]({{< relref "/api/credentials#get-credentials" >}}) operation in order to retrieve the
 [*x509-cert* type credentials]({{< relref "/api/credentials#x509-certificate" >}}) that are on record for the device.
+
+### JSON Web Token based Authentication
```

Comment:
Do you see this as a generic mechanism, i.e. not specific to Google IoT Core?

Reply:
Yes, this is definitively a generic mechanism. The goal was to minimize the migration effort for people coming from Google IoT Core while still keeping the mechanism generic.

Reply:
Well, the way it is described at the moment, i.e. retrieving the tenant and auth ID from the MQTT client identifier actually looks more like a Google IoT Core specific mechanism ...

Reply:
yeah, I think you are right. But it should definitively be a generic mechanism that also works for former Google IoT Core customers.
What is your opinion on the following compromise:
We add the possibility to provide the tenant-id and auth-id within the JWT claims (e.g. auth-id in `iss` and tenant-id in `aud`) and additionally keep the current approach (tenant-id and auth-id from client identifier) to make it easy for people to migrate from Google IoT Core. This should be pretty generic and also work for a possible extension to other protocol adapter (e.g. html) in the future.

Reply:
That is what I had in mind as well and hoped for you to propose ;-) We should, however, spend some more thought on which claims to expect to find the tenant and auth ID in ...

Reply:
An interesting aspect of this JWT based scheme is that the issuer is actually the same entity as the subject because the device creates the JWT asserting claims about *itself*, including its identity. However, you might argue that the device's identity is being asserted in the context of a tenant. In order to properly look up the device's public key on record in the device registry, we also need the tenant that the device belongs to. Thus, IMHO we should put the tenant ID into the *iss* claim and the authentication ID into the *sub* claim.

The audience of the JWT, determined by the value of the *aud* claim, should be something like `hono-adapter` in order to express that the JWT's purpose is to authenticate devices with Hono's protocol adapters.

Reply:
That's right, this is why I suggested to have the auth-id in either the `iss` or `sub` claim, as the device is technically both. But I really like your suggestion, so I will go with that :)

Reply:
I finished implementing the alternative approach like we talked about. If this approach is chosen the tenant-id and auth-id have to be provided within the `iss`  and `sub` claims, respectively, and the `aud` claim has to be "hono-adapter".

Reply:
I have given this some more thought. If we later also want to support a more generic JWT based authentication mechanism, that is, one that is not specific to Google IoT Core, then I believe we should allow for the `iss` claim to be used in its original way, i.e. to convey a URI that identifies the issuer of the token other than the device itself.
For example, if a third party had issued the token, then the `iss` claim might indicate an issuer that is trusted at the level of the tenant that the device belongs to. In order for the protocol adapter to validate the token, it would make sense to look up the public key within the tenant configuration instead of the device credentials. Instead of relying on URI parsing, I would suggest to have the protocol adapter check if the value of the `iss` claim is the same as the value of the `sub` claim (or empty). If so, the public key is to be looked up from the device's credentials on record. Otherwise, the value is being interpreted as a URI that can be used to find the public key within the tenant configuration. Note that we can focus on the former case in this PR. However, we should make sure that the latter can be easily added later on.
For this to work, we would need to put the tenant ID into a different claim. Something like `tid` comes to mind. WDYT?

Reply:
We think this would definitively make sense but would suggest to implement this in a future PR altogether. Does this work for you?